### PR TITLE
add "MultiSmoothSuperellipsoidIC" to phasefield module

### PR DESCRIPTION
**Pull request by FHilty**
### Complete each of the following items before creating a PR
- Include any relevant design information for your change
- Follow [Coding Standards](http://mooseframework.org/wiki/CodeStandards/)
- Submit one or more [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/) or modify existing test cases
- Reference or close a specific issue (refs # or closes #)
  …e creation of randomly located superellipsoids which can have either normal or uniformly distributed variation in the semiaxes of the ellipse. #7208

### CODE REVIEW

**Comment by permcody** [comment]
Please submit a test case for this new code.

Thanks!

File: modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  for (unsigned int i = 0; i < LIBMESH_DIM; i++)
+  {
+    _bottom_left(i) = _mesh.getMinInDimension(i);
+    _top_right(i) = _mesh.getMaxInDimension(i);
+  }
+  _range = _top_right - _bottom_left;
+
+  switch (_semiaxis_variation_type)
+  {
+  case 2: //No variation
```
Comment (FHilty) [inline modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C]: Thanks for drawing my attention to that piece of code. I fixed it to longer use a switch, also i noticed that as it was written the breaks would have prevented it from working as intended anyway.

Comment (FHilty) [inline modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C]: Test case created and amended

File: modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+      {
+        if (j == 0) rr = _range.norm();
+
+        Real tmp_rr = _mesh.minPeriodicDistance(_var.number(), _centers[j], newcenter);
+
+        if (tmp_rr < rr) rr = tmp_rr;
+      }
+
+      if (i == 0) rr = _range.norm();
+    }
```
Comment (permcody) [inline modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C]: interesting... It seems like this code would be useful in other ICs. We have the same problem with assigning centers in the VoronoiIC as well. Are you interested in putting this code into the PhaseField utils? If not that's fine. somebody else can move it later.

File: modules/phase_field/tests/initial_conditions/MultiSmoothSuperellipsoidIC_Test.i [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  [../]
+[]
+
+[Executioner]
+  # Preconditioned JFNK (default)
+  type = Transient
+  scheme = bdf2
+  solve_type = PJFNK
+  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart -mat_mffd_type'
+  petsc_options_value = 'hypre boomeramg 101 ds'
```
Comment (permcody) [inline modules/phase_field/tests/initial_conditions/MultiSmoothSuperellipsoidIC_Test.i]: Do you need a 101 restart vectors? i.e. are you really taking that many linear its or is this copy and pasted from some other file?

File: modules/phase_field/tests/initial_conditions/MultiSmoothSuperellipsoidIC_Test.i [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  l_tol = 1e-4
+  nl_max_its = 20
+  nl_rel_tol = 1e-9
+  nl_abs_tol = 1e-11
+  start_time = 0.0
+  num_steps = 1
+  dt = 100.0
+  [./Adaptivity]
+    refine_fraction = .5
+  [../]
```
Comment (permcody) [inline modules/phase_field/tests/initial_conditions/MultiSmoothSuperellipsoidIC_Test.i]: This is a pretty hefty test! My guess is that this doesn't run in under 2 seconds or even in that ballpark. How about in full debug mode on Linux? I guess we'll see soon enough.

File: modules/phase_field/tests/initial_conditions/MultiSmoothSuperellipsoidIC_Test.i (hunk 3 above)
Comment (FHilty) [inline modules/phase_field/tests/initial_conditions/MultiSmoothSuperellipsoidIC_Test.i]: This was copy and pasted from phase_field/tests/MultiSmoothCircleIC/multismoothcircleIC_test.i, i just changed IC from circles to ellipses.

File: modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C (hunk 2 above)
Comment (FHilty) [inline modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C]: Someone else should probably do it, i am still new and don't know how i would add this to PhaseField utils.

**Comment by tonkmr** [comment]
Since you are only testing an initial condition here, there isn't really any need to do a solve. You can just output the initial condition by adding this block:
```
[Problem]
solve = false
[]
```
You can remove all the kernels and everything. Since you are only creating the initial condition, it runs very fast so you could even test this in 3D (maybe have one 2D test and one 3D test).

For an example, see ```phase_field/tests/reconstruction/2phase_reconstruction3.i```

**Comment by permcody** [comment]
Good suggestion!

**Comment by FHilty** [comment]
Thanks for the advice Dr. Tonks, I removed the old test and created a 2D and 3D test case that completes much faster than the original.

**Comment by permcody** [comment]
@laagesen - Some of this good is similar to the PR you have up. Should the "centers" computation be abstracted out? Do all of these classes need a common base?

**Comment by laagesen** [comment]
Well this PR is doing the centers calculation sort of different than the way I am in the BimodalSuperellipsoids classes. @FHilty is checking the distance between centers and making sure that the distance between centers is not < `bubspac` as acceptance criteria for a new randomly placed bubble. In the `BimodalSuperellipsoids` classes, as an acceptance criteria, I'm checking to make sure the distance between the particles/bubbles along the line between their centers is above a certain value `small_spac` or `large_spac` depending whether it's a large or small bubble. Of course the two ways are equivalent for circular particles, but for my use case, I have nearly square larger particles, and there is a precipitate-free zone inside or outside of the larger particles, so I needed the boundary of the precipitate-free zone to follow the contour the particle, which isn't possible if you check only center-to-center distance. 

Overall, though, I agree that there is some stuff that could be shared between this and the `BimodalSuperellipsoids` ICs. I could create a function that calculates the distance between edges into a base class, and @FHilty (and I) could use that, although it would necessitate changing from checking center-to-center distances to checking edge-to-edge distance for @FHilty. By doing this you can guarantee that you would avoid having overlapping particles (although I'm assuming here you want to avoid having overlapping particles, maybe this is not the case!) However, to create the list of random center positions, for the `Bimodal` case you need to know about the large particles first so you don't put a small particle on top of them, so I don't think you can create the list of the randomly positioned particles in the same way for both cases. I think we could probably standardize the way we create the lists of exponents and semiaxes and put them in a base class too.

**Comment by FHilty** [comment]
Spooky. Today i started writing code to accomplish just what your saying @laagesen, i need to use 4-5 instances of MultiSmoothSuperellipsoidIC to re-create a specific microstructure. I was going to re-purpose code from BimodalSuperellipsoidsIC to check edge-to-edge between new centers and a list of accepted centers. It wouldn't completely prevent overlap if they were close, had significantly different aspect ratios and were aligned off-axis to each other, but that shouldn't be an issue for me. 

To keep track of centers, semiaxes, and exponents i wanted to store them in vectors and use push_back to amend newly accepted centers to the master list. The biggest unknown for me is if there is a way to enforce the order ICs are executed. If so i hoped to add a priority to MultiSmoothSuperellipsoidIC then If priority 1 it goes first and can initialize the master lists and so on until done. But i have no idea this is even a plausible plan, i only have about 2 weeks experience with object oriented coding yet.

**Comment by permcody** [comment]
There are other cases as well, but they are even trickier. I think we can do something with these centers to figure out the Voronoi IC too. This case is much trickier because it's not about how close the centers are, it's more about whether or not the spacing between several centers is far enough apart based on your mesh resolution and interface width to create valid Voronoi cells. This is important when running several simulations with a specific number of grains that begin from random Voronoi cell centers.

I'll review both of these PRs and get them merged and then perhaps you two can take a look at abstracting out common code where we can do that. I'm also willing to take a look at this too.

Tagging @dschwen and @tonkmr for additional comments.

File: modules/phase_field/include/ics/MultiSmoothSuperellipsoidIC.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef MULTISMOOTHSUPERELLIPSOIDIC_H
+#define MULTISMOOTHSUPERELLIPSOIDIC_H
+
+#include "Kernel.h"
+#include "SmoothSuperellipsoidBaseIC.h"
+
+// System includes
+#include <string>
```
Comment (permcody) [inline modules/phase_field/include/ics/MultiSmoothSuperellipsoidIC.h]: <string> isn't needed

File: modules/phase_field/include/ics/MultiSmoothSuperellipsoidIC.h [hunk 6]
```diff
@@ -0,0 +1,63 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef MULTISMOOTHSUPERELLIPSOIDIC_H
+#define MULTISMOOTHSUPERELLIPSOIDIC_H
+
+#include "Kernel.h"
```
Comment (permcody) [inline modules/phase_field/include/ics/MultiSmoothSuperellipsoidIC.h]: "Kernel" isn't needed

File: modules/phase_field/tests/initial_conditions/tests [hunk 7]
```diff
@@ -128,4 +128,18 @@
...
+    input = 'MultiSmoothSuperellipsoidIC_3D.i'
+    exodiff = 'MultiSmoothSuperellipsoidIC_3D_out.e'
+    scale_refine = 1
+  [../]
+
+  [./MultiSmoothSuperellipsoidIC_2D]
+    type = Exodiff
+    input = 'MultiSmoothSuperellipsoidIC_2D.i'
+    exodiff = 'MultiSmoothSuperellipsoidIC_2D_out.e'
+    scale_refine = 1
```
Comment (permcody) [inline modules/phase_field/tests/initial_conditions/tests]: Please drop theses scale_refine options from your tests file (both).

**Comment by permcody** [comment]
When you push your next set of changes. Feel free to delete the PR template in the description box.

**Comment by permcody** [comment]
👍

**Pull request merged**